lib/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.cpp

```diff
@@ -125,47 +122,80 @@ DIBasedTypeHierarchy::DIBasedTypeHierarchy(const LLVMProjectIRDB &IRDB) {
     TransitiveClosure[I][I] = true;
   }
 
-  // get VTables
-  const auto Subprograms = Finder.subprograms();
-  for (auto *Subprogram : Subprograms) {
+  size_t VTableSize = 0;
+  for (const auto &Subprogram : Finder.subprograms()) {
+    if (const auto *SubProgramType =
+            llvm::dyn_cast<llvm::DISubprogram>(Subprogram)) {
+      if (SubProgramType->getVirtualIndex() > VTableSize) {
+        VTableSize = SubProgramType->getVirtualIndex();
+      }
+    }
+  }
+  // if the biggest virtual index is 2 for example, the vector needs to have a
+  // size of 3 (Indices: 0, 1, 2)
+  VTableSize++;
+
+  std::vector<const llvm::Function *> Init = {};
+  std::vector<std::vector<const llvm::Function *>> IndexToFunctions;
+  for (size_t I = 0; I < VTableSize; I++) {
+    IndexToFunctions.push_back(Init);
+  }
 
+  // get VTables
+  for (auto *Subprogram : Finder.subprograms()) {
     if (const auto *SubProgramType =
             llvm::dyn_cast<llvm::DISubprogram>(Subprogram)) {
       if (!SubProgramType->getVirtuality()) {
         continue;
       }
-      // get all virtual functions
-      std::vector<const llvm::Function *> VirtualFunctions;
-      for (const auto &Function : Module->functions()) {
-        if (SubProgramType->getLinkageName() ==
-            Function.stripPointerCasts()->getName().str()) {
-          VirtualFunctions.push_back(&Function);
-        }
+
+      const auto *const FunctionToAdd =
+          IRDB.getFunction(SubProgramType->getLinkageName());
+
+      if (!FunctionToAdd) {
+        continue;
       }
-      VTables.push_back(LLVMVFTable(VirtualFunctions));
+
+      const size_t VirtualIndex = SubProgramType->getVirtualIndex();
+      std::vector<const llvm::Function *> IndexFunctions;
+
+      IndexFunctions.push_back(FunctionToAdd);
+      // concatenate vectors of functions of this index
+      IndexToFunctions[VirtualIndex].insert(
```

Comment:
As pointed out above, we don't want to index into the outer vector by VirtualIndex. Instead we want to find out which type this function belongs to and use the type's index instead